### ISSUE TITLE
More strict data type for member roles

### ISSUE THREAD

**Issue by taufik-rama** [serenity-rs/serenity#2097]
Hello :wave: 

I've been debugging an issue, it seems to be related to how roles are handled.

```console
$ cargo run
bot is connected!

# ...

Caused by:
    Invalid Form Body (roles.1: The set already contains this value)

# ...
```

Based on the message, I'd assume that there's at least a requirement that when we add a role to a server member, the list of roles shouldn't be duplicated. If this is the case, we could change the [`roles`](
https://docs.rs/serenity/latest/src/serenity/model/guild/member.rs.html#40) field data type into a [`HashSet`](https://doc.rust-lang.org/stable/std/collections/struct.HashSet.html)

```rust
/// Information about a member of a guild.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[non_exhaustive]
pub struct Member {
    /// ...

    /// Vector of Ids of [`Role`]s given to the member.
    pub roles: Vec<RoleId>, // Change into `HashSet<RoleId>`
    
    /// ...
```

This ensures that the roles can never be duplicated. Also to note that this would be an API break since it's a public item

I've taken a peek at other (internal implementation) location, and I think changing the type shouldn't be very difficult, since the operation on `Vec` should be able to applied also to `HashSet`

As a context: the way I handle the roles assignment also uses `HashSet`, and currently I just convert it to `Vec` specifically for this field

**Comment by kangalio** [comment]
![Screenshot_20220910_000220](https://user-images.githubusercontent.com/21220820/189451817-e253aebb-f3c0-4af7-9573-9cb82130eeaa.png)

There are a lot of `Vec<Id>`. HashSet has higher overhead (2x stack size, 3x slower in ad-hoc benchmark), which makes me reluctant to change all those to HashSet. Do you think the higher overhead is worth it? Or is it ok to be inconsistent and change just `Member::roles` to HashSet?

**Comment by taufik-rama** [comment]
Ah, I didn't realize there are more fields with the same pattern

To be clear, the root cause for what I'm trying to raise is that a duplicated values will result in an error when the Discord's API gets called. If the API can tolerate duplicates, it might be fine to just keep it with `Vec`s.

I obviously don't have the data on how people uses the library, but for my usecase (web devs) generally the performance hit is negligible compared to time spent waiting on I/O. Not to mention that this fields usually only gets processed once in a while. If I can raise a point, I'd say that `Vec` is _fast_, rather than `HashSet` being the one that is slow :p

tl;dr: it might be more proper to change the fields that better correspond to what Discord APIs expects

Having said that, I understand if this changes might not be considered, since it will break existing Serenity APIs since it's a publicly accessible fields

**Comment by kangalio** [comment]
Any others' opinions? Use `HashSet` for all sets of non-duplicating IDs, which prevents duplicates before reaching Discord's servers but increases struct sizes?

**Comment by taufik-rama** [comment]
Just that from me :+1:

**Comment by ivinjabraham** [comment]
Using `HashSet` instead of `Vec`s might be the most straightforward approach. Any deduping or validation logic intuitively seems to be a worse hit on average performance. 

We could use a custom generic wrapper around `Vec<T>` that implements the same API except changing insertion methods to check for duplicates and change all the fields to use the wrapper instead. But this is a lot more convoluted than just rewriting the API to use `Hashset`; it might be a bit more trouble upfront but should make the code a lot cleaner eventually.

**Comment by jamesbt365** [comment]
This is a massive memory overhead that just isn't worth it for an issue that stems from the developer not checking if a user has a role before attempting to add one.

Changing our API to use `HashSet` wouldn't even fix this as you are sending a request to discord to *add* roles that are already applied to the user, the real solution if developers don't want to send requests unnecessarily or get hit with a failed request would be to check against the existing user themselves.
